engine/gamesys/src/gamesys/scripts/script_resource.cpp

```diff
@@ -460,6 +461,241 @@ static int GraphicsTextureTypeToImageType(int texturetype)
+ *    go.set("#model", "texture0", my_texture_id)
+ * end
+ * ```
+ */
+static int CreateTexture(lua_State* L)
+{
+    // This function pushes the hash of the resource created
+    int top                  = lua_gettop(L);
+    const char* path_str     = luaL_checkstring(L, 1);
+    const char* texturec_ext = ".texturec";
+
+    char buf_ext[64];
+    const char* path_ext = dmResource::GetExtFromPath(path_str, buf_ext, sizeof(buf_ext));
+
+    if (path_ext == 0x0 || dmStrCaseCmp(path_ext, texturec_ext) != 0)
+    {
+        luaL_error(L, "Unable to create texture, path '%s' must have the %s extension", path_str, texturec_ext);
+        return 0;
+    }
+
+    char canonical_path[dmResource::RESOURCE_PATH_MAX];
+    uint32_t canonical_path_len  = dmResource::GetCanonicalPath(path_str, canonical_path);
+    dmhash_t canonical_path_hash = dmHashBuffer64(canonical_path, canonical_path_len);
+
+    if (dmResource::FindByHash(g_ResourceModule.m_Factory, canonical_path_hash))
+    {
+        luaL_error(L, "Unable to create texture, a resource is already registered at path '%s'", path_str);
+        return 0;
+    }
+
+    dmGameObject::HInstance sender_instance = dmScript::CheckGOInstance(L);
+    dmGameObject::HCollection collection    = dmGameObject::GetCollection(sender_instance);
+
+    luaL_checktype(L, 2, LUA_TTABLE);
+    uint32_t type        = (uint32_t) CheckTableInteger(L, 2, "type");
+    uint32_t width       = (uint32_t) CheckTableInteger(L, 2, "width");
+    uint32_t height      = (uint32_t) CheckTableInteger(L, 2, "height");
+    uint32_t format      = (uint32_t) CheckTableInteger(L, 2, "format");
+    uint32_t max_mipmaps = (uint32_t) CheckTableInteger(L, 2, "max_mipmaps", 0);
+
+    uint8_t max_mipmaps_actual = dmGraphics::GetMipmapCount(dmMath::Max(width, height));
+
+    if (max_mipmaps > max_mipmaps_actual)
+    {
+        dmLogWarning("Max mipmaps %d requested for texture %s, but max mipmaps supported for size (%d, %d) is %d",
+            max_mipmaps, path_str, width, height, max_mipmaps_actual);
+        max_mipmaps = max_mipmaps_actual;
+    }
+
+    // Max mipmap count is inclusive, so need at least 1
+    max_mipmaps                                        = dmMath::Max((uint32_t) 1, max_mipmaps);
+    uint32_t tex_bpp                                   = dmGraphics::GetTextureFormatBitsPerPixel((dmGraphics::TextureFormat) format);
+    dmGraphics::TextureImage::Type tex_type            = (dmGraphics::TextureImage::Type) GraphicsTextureTypeToImageType(type);
+    dmGraphics::TextureImage::TextureFormat tex_format = (dmGraphics::TextureImage::TextureFormat) GraphicsTextureFormatToImageFormat(format);
+    dmGraphics::TextureImage texture_image             = MakeTextureImage(width, height, max_mipmaps, tex_bpp, tex_type, tex_format);
+
+    dmMutex::HMutex mutex = dmResource::GetLoadMutex(g_ResourceModule.m_Factory);
+    while(!dmMutex::TryLock(mutex))
+    {
+        dmTime::Sleep(100);
+    }
```

Comment:
I think this needs commenting.

Why polling the state here?
Why not use a Lock/Unlock?

Also, we're already on the main thread, and afaik, the resource loading happens on the main thread too?

Also, a follow up question: What happens if I create a new dynamic texture in the "init" flow of a script? I guess you've testes, since that's perhaps the most natural place to put it (as opposed to creating it in the update loop).
The reason I ask, is because the init flow of the script is likely within the resource creation phase of the collection/gameobject, and might also touch the resource mutex.